### PR TITLE
Fix CsaGuzzleExtension to allow symfony services

### PR DESCRIPTION
**Pull request by Khez**
Hello,
This is a fix in CsaGuzzleExtension to allow symfony services where class instances are used directly from config array.

In GuzzleHttp/Client::__construct(), four keys are expected to be instances of a class.
Namely *message_factory*, *fsm*, *handler* and *adapter*.

My personal use case, required me to extend the _message_factory_ class for some special routines.
This appears to be an oversight in both 1.3 and 2.0 (where only the handler key remains)

Also added some tests for this scenario.

Best Regards,
Khez

### CODE REVIEW

**Comment by csarrazi** [comment]
Oh, great point! I was considering adding this for `2.0` (and actually, in `2.0`, the handler is indeed defined, and injected as a service, as the only means to inject middleware is through the handler stack).

File: src/Tests/DependencyInjection/CsaGuzzleExtensionTest.php [hunk 1]
```diff
@@ -64,6 +64,32 @@ class: AppBundle\Client
         $this->assertEquals('AppBundle\Client', $client->getClass());
     }
 
+    /**
+     * @dataProvider clientConfigInstance
+     * @covers CsaGuzzleExtension::buildGuzzleConfig
+     */
+    public function testClientConfigInstanceOverride($instanceKey)
+    {
+        $serviceId = 'my.message.factory.id';
```
Comment (csarrazi) [inline src/Tests/DependencyInjection/CsaGuzzleExtensionTest.php]: Why not inject both the service id and the instance key in the data provider?

File: src/DependencyInjection/CsaGuzzleExtension.php [hunk 2]
```diff
@@ -132,4 +133,20 @@ function ($tags) { return $tags[0]['alias']; },
             return !isset($explicitlyConfiguredIds[$name]) || $explicitlyConfiguredIds[$name];
         });
     }
+
+    private function buildGuzzleConfig($options)
+    {
+        if (!isset($options['config'])) {
+            return null;
```
Comment (csarrazi) [inline src/DependencyInjection/CsaGuzzleExtension.php]: You can omit null here. Simply use
 
```php
if (!isset($options['config'])) {
    return;
}
```

Comment (csarrazi) [inline src/DependencyInjection/CsaGuzzleExtension.php]: Also, I would actually move this part up in the previous method (`processClientsConfiguration()`), in order to have less code, and directly inject the `$config` in the `buildGuzzleConfig()` method.

i.e.

```php
if (isset($options['config'])) {
    $client->replaceArgument(0, $this->buildGuzzleConfig($options['config']));
}

// ...

public function buildGuzzleConfig(array $config)
{
    // ...
}
```

This way, you can also ensure that `$config` key is an array.

**Comment by csarrazi** [comment]
Apart from my very few comments, I'm :+1: ! :smiley:

File: src/Tests/DependencyInjection/CsaGuzzleExtensionTest.php (hunk 1 above)
Comment (Khez) [inline src/Tests/DependencyInjection/CsaGuzzleExtensionTest.php]: Since we're not actually testing the serviceId in a meaningful way, didn't feel the need :?

If it's a deal breaker, I can update the PR :)

File: src/DependencyInjection/CsaGuzzleExtension.php (hunk 2 above)
Comment (Khez) [inline src/DependencyInjection/CsaGuzzleExtension.php]: Valid :+1:

File: src/Tests/DependencyInjection/CsaGuzzleExtensionTest.php (hunk 1 above)
Comment (csarrazi) [inline src/Tests/DependencyInjection/CsaGuzzleExtensionTest.php]: Nah, it was just an open question, don't bother with this! ;)

**Comment by Khez** [comment]
Hello,

Tweaked the code as per your feedback.

Also added a new test for non-array config (should never happen?)
Possibly this would be better suited for the Configuration class...

Best Regards,
Khez

**Comment by csarrazi** [comment]
Perfect! Could you simply add a test to make sure that injecting `null` in the client's configuration doesn't break the configuration?

e.g.

```yml
csa_guzzle:
    clients:
        github:
            config: ~
```

I'll merge your PR right after! :)

**Comment by Khez** [comment]
Done!

For a moment I thought I might have created a BC break... looks like I didn't :D

Good catch!

**Comment by csarrazi** [comment]
Great work @Khez! :+1:
